### PR TITLE
User login

### PR DESCRIPTION
**Pull request by mbodeantor**
Adding password reset endpoint, splitting out queries into functions for unit tests, splitting out API key generation to its own endpoint

### To test login:
1. signing up and / or logging in via the /login route
To test: navigate to /login and perform whichever operation. (Sign up a user, or you can use test-022324@domain.com | Password1! to log in)

2. Gating protected routes.
To test: If not logged in, navigate to /change-password (a blank page presently), and you should be redirected to /login. Log in, and you should be redirected back to /change-password.

3. Automatically refresh user's access tokens if they are active within 1 minute of current token expiration.
To test: Log in with a user. Wait 4 minutes. Click, press keys, or scroll anywhere in the UI (doesn't have to be a button or anything). Wait another minute and navigate to /change-password. You should not be redirected back to /login.

4. Automatically log a user out on user action when the access token has expired
To test: Log in with a user and navigate to /change-password. Wait 5+ minutes and click/press/scroll anywhere in the UI. You should be redirected to /login (if not on a protected route, there will be no navigation, so it won't interrupt users who are using /, /search, etc.

### To test changing password:
1. Go to `/login` and log in with any user.
2. Navigate to `/change-password`
3. Change your password—you will be automatically logged back in with the new password.

### To test reseting password
1. Sign up a user with a valid email address (you can use [Ethereal](https://ethereal.email/) to create a testing account)
2. Log out
3. From `/login`, click `Click here to reset it`
4. Enter the email you used to sign up.
5. Follow the link that is delivered in the email.
6. Change password
7. After success, click `Click here to log in`
8. Confirm that the updated password logs your user in.

### CODE REVIEW

**Comment by josh-chamberlain** [comment]
@mbodeantor nice! I think this needs a docs update and ideally `steps to test` before I would be able to review it, from a technical skills perspective

**Review by kalenluciano — APPROVED** [review]
This looks good to me!

**Comment by mbodeantor** [comment]
@josh-chamberlain Docs are updated here: https://app.gitbook.com/o/-MXypK5ySzExtEzQU6se/s/-MXyolqTg_voOhFyAcr-/~/changes/464/api/endpoints/admin

Here are the new tests I added for the new endpoints to try them out:
```
def test_put_user():
    response = requests.get(
        "https://data-sources.pdap.io/api/user",
        headers=HEADERS,
        json={"email": "test2", "password": "test"},
    )

    return response.json()["data"] == "Successfully updated password"


# api-key
def test_get_api_key():
    response = requests.get(
        "https://data-sources.pdap.io/api/api_key",
        headers=HEADERS,
        json={"email": "test2", "password": "test"},
    )

    return len(response.json()["api_key"]) > 0
```

**Comment by mbodeantor** [comment]
Put this on hold so the front end work can be added here

File: resources/User.py
Comment (joshuagraber) [inline resources/User.py:53]: @mbodeantor In the case of resetting a user's password, how are we authenticating them? Given the structure of this endpoint, anyone with access could change anyone else's password.

I'm thinking we might want to send the user an email with a magic link, and only by clicking that email are they able to access the frontend UI for changing a password. Wyt?

**Review by joshuagraber — COMMENTED** [review]
I've started looking into the client work for this @mbodeantor and just have a question for you on the password reset endpoint.

File: resources/User.py
Comment (mbodeantor) [reply resources/User.py:53]: @joshuagraber That makes sense to me. What API changes would be required for that?

Comment (mbodeantor) [reply resources/User.py:53]: I'm on board except I think instead of assuming in point 2, can we have a "forgot password?" link that only requires them to send an email?

Comment (josh-chamberlain) [reply resources/User.py:53]: I still think @mbodeantor should answer, but I'll note here that we could consider adding an external library so we don't reinvent any wheels:

- [flask-security-too](https://flask-security-too.readthedocs.io/)
- [authlib](https://authlib.org/)
- [flask-user](https://flask-user.readthedocs.io/en/latest/) + [flask-login](https://pypi.org/project/Flask-Login/)

Comment (joshuagraber) [reply resources/User.py:53]: For sure @mbodeantor I'm thinking that there can be a "change password" link (enter current pw + desired updated pw) and a "reset password" link (only email). My thinking above was if we were using the same endpoint for both, the API would need to perform some logic there.

**Comment by joshuagraber** [comment]
Started working on the basic UI for this. @mbodeantor are you good if I do the client work in this branch / PR as well?

**Comment by joshuagraber** [comment]
@mbodeantor Needed to make [a few updates to `design-system` to properly accomplish the client work for this](https://github.com/Police-Data-Accessibility-Project/design-system/pull/50). Symlinked locally and added a screen recording, but if you want to test more thoroughly, I can push some testing code to this branch.

File: resources/User.py
Comment (joshuagraber) [reply resources/User.py:53]: Perfect timing, @mbodeantor I've got almost everything else ready to go locally for the client stuff.

Comment (mbodeantor) [reply resources/User.py:53]: @josh-chamberlain Looks like we will need an email server in order to properly send these password reset emails. Do we already have one by chance? If not, looks like Sendgrid has a limited free plan:
https://sendgrid.com/en-us/marketing/sendgrid-services-cro#pricing

Comment (joshuagraber) [reply resources/User.py:53]: @mbodeantor @josh-chamberlain Doesn't python have built-in support for sending smtp emails? Presumably we already have email set up for pdap.io, couldn't we just set up a no-reply address and send using that? (Sendgrid is great, though. Have used previously for node projects, where there is no native email functionality.)

**Comment by joshuagraber** [comment]
One other thing I'm wondering here @josh-chamberlain @mbodeantor (a next-week question 😅 ): how are we handling sessions? Are we handling this entirely client-side? Or do we want to use an access-token solution?

FWIW, I vote for the latter, as it's much more secure (particularly since we'll eventually be adding the ability for users to modify records in the db).

**Comment by josh-chamberlain** [comment]
@joshuagraber @mbodeantor I advocate for the latter, too—it's what we've talked about, and supported by auth libraries if needed

File: resources/User.py
Comment (mbodeantor) [reply resources/User.py:53]: @josh-chamberlain @joshuagraber An email server will allow our password reset emails to be successfully delivered, emails from a random python app will likely all be blocked

**Comment by joshuagraber** [comment]
> and supported by auth libraries if needed

@josh-chamberlain I think it makes sense to use a trusted lib rather than building ourselves, but I defer to @mbodeantor on that.

I pushed the initial `/login` client work. It's testable. I confirmed by signing up with a new user and logging in with that user.
I'll pause the client work for now, though, because depending on how we implement the token stuff, it may change that approach and the approach for the password reset stuff.

**Comment by mbodeantor** [comment]
I tried to standardize responses from all endpoints to contain a "message" key which will contain the error if one is returned. Successful responses will have a "data" keyword. Didn't seem to break anything, but wanted to make sure to call it out

**Comment by mbodeantor** [comment]
Will also now need to export the MAILGUN_KEY variable for the app to be able to send password reset emails. Please see the DigitalOcean app variables for the key or DM me.

**Comment by mbodeantor** [comment]
/login is now its own endpoint. Docs: https://app.gitbook.com/o/-MXypK5ySzExtEzQU6se/s/-MXyolqTg_voOhFyAcr-/~/changes/471/api/endpoints/admin

File: resources/Login.py
Comment (joshuagraber) [inline resources/Login.py:30]: Hey @mbodeantor Quick question here: Is this API key functioning as an access token for authenticating sessions or is it a unique key per user that is stable for accessing the API?
I don't think we want to use stable API keys as access tokens. Could we use a JWT instead? Could very easily encode an expiry time into the token so that we can use a refresh token to generate a new access token when a user's access token is about to expire and they're still using the app. I believe this is the strategy we discussed previously with @josh-chamberlain, no?

Comment (mbodeantor) [reply resources/Login.py]: should be fixed now

**Comment by mbodeantor** [comment]
@joshuagraber should be g2g

File: resources/User.py
Comment (joshuagraber) [inline resources/User.py:24]: One final thing here @mbodeantor, can we call the login endpoint from here to automatically log a user in once they've signed up? Or maybe as the API is concerned, all that means is to call the token middleware and return the token? I think it would be nice to automatically begin a session for new users rather than making them enter the same info to log in after signing up.

Comment (joshuagraber) [reply resources/User.py:24]: Another option here is for the client to just automatically hit the login endpoint after signup. I think either way works, just lmk what you'd prefer.

Comment (mbodeantor) [reply resources/User.py:24]: Yeah I think doing it automatically client side might make more sense. Not sure how common it is but like Max will be creating his own user from the API and won't need the login.

Comment (joshuagraber) [reply resources/User.py:24]: Ah, right, that makes sense. And it keeps the API separated.

**Comment by joshuagraber** [comment]
Hey @josh-chamberlain I've updated the login stuff, with @mbodeantor's help, over the course of the day today. Didn't get to the password change/reset work yet, will handle next week.

The login work takes care of the following, if you're interested in testing stuff out.

1. signing up and / or logging in via the `/login` route
To test: navigate to `/login` and perform whichever operation. (Sign up a user, or you can use `test-022324@domain.com` | `Password1!` to log in)

2. Protecting authenticated routes.
To test: If _not_ logged in, navigate to `/change-password` (a blank page presently), and you should be redirected to `/login`. Log in, and you should be redirected back to `/change-password`.

3. Automatically refresh user's access tokens if they are active within 1 minute of current token expiration.
To test: Log in with a user. Wait 4 minutes. Click, press keys, or scroll anywhere in the UI (doesn't have to be a button or anything). Wait another minute and navigate to `/change-password`. You should be successfully redirected.

4. Automatically log a user out on user action when the access token has expired
To test: Log in with a user and navigate to `/change-password`. Wait 5+ minutes and click/press/scroll anywhere in the UI. You should be redirected to `/login` (if not on a protected route, there will be no navigation, so it won't interrupt users who are using `/`, `/search`, etc.

For ease of understanding the timing of access token stuff, I added a debug log to the console for when a user is logged in, which looks like this:
Screenshot 2024-02-23 at 2 06 15 PM

**Comment by josh-chamberlain** [comment]
thanks y'all! I started looking at this but I won't be able to take a proper look until next week.

File: resources/Login.py
Comment (joshuagraber) [inline resources/Login.py:23]: @mbodeantor Can we error here if the user does not exist in DB? Currently it's returning `200` with a `null` response.

File: resources/RequestResetPassword.py
Comment (joshuagraber) [inline resources/RequestResetPassword.py:15]: @mbodeantor This endpoint is failing with the following error:

`415 Unsupported Media Type: Did not attempt to load JSON data because the request Content-Type was not 'application/json'.`

The `Content-Type` header is indeed set to `"application/json"`, so I'm not sure what's happening here.

Screenshot 2024-02-26 at 9 36 15 AM

File: resources/User.py
Comment (joshuagraber) [inline resources/User.py:35]: @mbodeantor Since this is an authenticated route on the client (as opposed to the reset route, which is unauthenticated), can we check the access token here (passed as the `Authorization` header) before changing the PW?

**Review by joshuagraber — COMMENTED** [review]
Hey @mbodeantor, I did some work on the change/reset password stuff this morning.

A couple of questions / comments. TLDR: One inexplicable error around the reset stuff, and a security suggestion for an already-authenticated user changing pw.

File: resources/RequestResetPassword.py
Comment (joshuagraber) [reply resources/RequestResetPassword.py:15]: Update: hitting this from command line with the following, it works correctly. So I'm not sure what's different about this request from the browser compared with others?

```
curl -X GET -H "Content-Type: application/json" --data "{\"email\": \"mason.west38@ethereal.email\"}" http://localhost:5000/request-reset-password
```

File: resources/RequestResetPassword.py
Comment (joshuagraber) [inline resources/RequestResetPassword.py:26]: Oh, one other thing here. We want the base url of the client. This is sending the base url of the API.

Comment (mbodeantor) [reply resources/RequestResetPassword.py:26]: This will require an additional env variable VITE_VUE_API_BASE_URL which will have to replace all current instances of VITE_VUE_APP_BASE_URL in the client code. They will be the same locally but will need to be updated in DO when these are merged

**Review by josh-chamberlain — CHANGES_REQUESTED** [review]
This looks good! We can circle back for the tests and a couple other enhancements.

For now, could we just change two things?

- it would be great to rebase this to merge into dev first, and test it there before merging dev to main. 

- instead of "Don't have an account? Sign up" and "Forgot your password? Click here to reset it", let's just have  two secondary buttons (mocked up here)
Screen Shot 2024-02-26 at 3 21 41 PM

**Comment by joshuagraber** [comment]
@josh-chamberlain Thank you! I've made the styling updates here.

 As far as rebasing onto `dev` goes, that would be pretty messy right now, because  1. this is a giant PR, and 2. `dev` is currently pretty far out of date from `main`. I agree, though, that this would be nice to test on `dev` first, particularly since we added new env vars that the whole app will depend on. How far are we from a point where we could bring `dev` and `main` into parity?  cc @mbodeantor

**Comment by mbodeantor** [comment]
@josh-chamberlain @joshuagraber I just forced a rebuild to make sure dev is working, I'll check tomorrow but everything on dev should go into main. After that merge I agree we should merge this branch into dev to make sure these large changes deploy correctly.

**Comment by joshuagraber** [comment]
Hey @mbodeantor I'm working on writing some tests for the change/reset password routes, and in the process I noticed that login is failing.

Request: 
```
{
  "email": "test-021224@domain.com",
  "password": "Password1!"
}
```

Response (`500`):
```
{
    "message": "Expected a string value"
}
```

**Comment by mbodeantor** [comment]
@joshuagraber can you verify SECRET_KEY is in your environment locally? I had this issue in GH tests but they're passing locally for me

**Comment by mbodeantor** [comment]
No, but it wasn't being used recently.

**Comment by joshuagraber** [comment]
> No, but it wasn't being used recently

So does this need to be passed to the API somewhere? It's exported via the shell alias, but the client app doesn't have access to it in code right now. 


> I had this issue in GH tests but they're passing locally for me

Are you able to run the client app locally and log in via the browser with your current configuration?

File: middleware/login_queries.py
Comment (joshuagraber) [inline middleware/login_queries.py:42]: @mbodeantor looks like we're looking for `secret_key` rather than `SECRET_KEY`. When I change this line, the client is fine locally with the uppercased value. 

```suggestion
    session_token = jwt.encode(payload, os.getenv("SECRET_KEY"), algorithm="HS256")
```

Do you prefer this var to be exported in snake case rather than screaming snake?

Comment (mbodeantor) [reply middleware/login_queries.py:42]: Ah no is should be uppercase, forgot to change it back after trying to get it to work in GH.

Comment (joshuagraber) [reply middleware/login_queries.py]: Alright, I committed it. 🙏

**Review by josh-chamberlain — APPROVED** [review]
definitely good enough to merge into dev!

**Pull request merged**
